### PR TITLE
AIP-44 Migrate BaseJob.run to Internal API

### LINKED ISSUE #29316: AIP-44 Migrate BaseJob.run to Internal API
https://github.com/apache/airflow/blob/main/airflow/jobs/base_job.py#L248

### PR DESCRIPTION
**Pull request by pierrejeambrun**
closes: https://github.com/apache/airflow/issues/29316

### CODE REVIEW

**Review by dimberman — APPROVED** [review]

**Comment by mhenc** [comment]
I don't think this is what we need to do there (sorry I was not more explicit in the bug description).

the `self.execute()` needs to be run locally (e.g on worker) as this is the part that touches customer code, your change will make it executed in Internal API sever.

What we need to do there is migrate everything that touches "session" - so the pre-execute (changing state etc) and post-execute - updateing state,end_date.

**Comment by pierrejeambrun** [comment]
My bad, I didn’t pay enough attention and assumed it was like other issues.

I will update the PR, thanks

**Comment by pierrejeambrun** [comment]
Do we have a way of actually testing the code, with `AIRFLOW__WEBSERVER__RUN_INTERNAL_API`. Last time I tried it wasn't ready but would be great to be able to run this locally with internal_api.

File: airflow/jobs/base_job.py [hunk 1]
```diff
@@ -245,32 +246,47 @@ def heartbeat(self, only_if_necessary: bool = False):
...
 
+    @staticmethod
+    @internal_api_call
+    @provide_session
+    def pre_execute(job: BaseJob, *, session: Session = NEW_SESSION) -> None:
+        """Update the database job entry before running the _execute() function."""
+        job.state = State.RUNNING
+        session.add(job)
+        session.commit()
+        make_transient(job)
```
Comment (pierrejeambrun) [inline airflow/jobs/base_job.py:257]: Wondering if this `make_transient` call is necessary in case of execution on the internal api.

Comment (potiuk) [reply airflow/jobs/base_job.py:257]: I think there is a bigger problem here. BaseJob is an SQLAlchemy model and we are effectively trying to serialize it with internal_api call. This is not going to end well.

I think this is one of those cases where we need to do it differently and refactor the code quite a bit deeper.

Comment (potiuk) [reply airflow/jobs/base_job.py:257]: (BTW. This is the first case I see we try to do that, so it's not a general issue so far).

**Review by potiuk — CHANGES_REQUESTED** [review]
Requesting changes so that it is not merged accidentally before the "BaseJob" passing is resolved.

File: airflow/jobs/base_job.py (hunk 1 above)
Comment (pierrejeambrun) [reply airflow/jobs/base_job.py:257]: Yep, much bigger problem :joy:.

Converting to draft as I will not be able to update before going on holidays.

**Comment by mhenc** [comment]
Regarding testing it: you may check
https://github.com/apache/airflow/pull/28900#issuecomment-1431391228

**Comment by potiuk** [comment]
not stale  I guess.

**Pull request closed without merging**